scripts/bert/fp16_utils.py

```diff
@@ -0,0 +1,231 @@
+
+        Parameters
+        ----------
+        batch_size : int
+            Batch size of data processed. Gradient will be normalized by `1/batch_size`.
+            Set this to 1 if you normalized loss manually with `loss = mean(loss)`.
+        max_norm : NDArray, optional, default is None
+            max value for global 2-norm of gradients.
+        """
+        self.fp32_trainer.allreduce_grads()
+        step_size = batch_size * self._scaler.loss_scale
+        if max_norm:
+            norm, ratio = grad_global_norm(self.fp32_trainer._params,
+                                           max_norm * self._scaler.loss_scale)
+            step_size = ratio * step_size
+            self.fp32_trainer.update(step_size)
+            overflow = not np.isfinite(norm.asscalar())
+        else:
+            # TODO(haibin) optimize the performance when max_norm is not present
+            self.fp32_trainer.update(step_size)
+            overflow = self._scaler.has_overflow(self.fp32_trainer._params)
+
+        # update scale based on overflow information
+        self._scaler.update_scale(overflow)
+        if overflow:
+            for param in self.fp32_trainer._params:
+                param.zero_grad()
+
+class LossScaler(object):
+    """Abstract loss scaler"""
+    def has_overflow(self, params):
+        """ detect inf and nan """
+        is_not_finite = 0
+        for param in params:
+            if param.grad_req != 'null':
+                grad = param.list_grad()[0]
+                is_not_finite += mx.nd.contrib.isnan(grad).sum()
+                is_not_finite += mx.nd.contrib.isinf(grad).sum()
+        # NDArray is implicitly converted to bool
+        if is_not_finite == 0:
+            return False
+        else:
+            return True
+
+class StaticLossScaler(LossScaler):
+    """Static loss scaler"""
+    def __init__(self, init_scale=1):
+        self.loss_scale = init_scale
+
+    def update_scale(self, overflow):
+        """update loss scale"""
+        pass
+
+class DynamicLossScaler(object):
+    """Class that manages dynamic loss scaling."""
+    def __init__(self, init_scale=2.**15, scale_factor=2., scale_window=2000,
+                 tolerance=0.05, verbose=False):
+        self.loss_scale = init_scale
+        self.scale_factor = scale_factor
+        self.scale_window = scale_window
+        self.tolerance = tolerance
+        self._num_steps = 0
+        self._last_overflow_iter = -1
+        self._last_rescale_iter = -1
+        self._overflows_since_rescale = 0
+        self._verbose = verbose
+
+    def update_scale(self, overflow):
+        """dynamically update loss scale"""
+        iter_since_rescale = self._num_steps - self._last_rescale_iter
+        if overflow:
+            self._last_overflow_iter = self._num_steps
+            self._overflows_since_rescale += 1
+            percentage = self._overflows_since_rescale / float(iter_since_rescale)
+            if percentage >= self.tolerance:
```

Comment:
Are you scaling only if > 5% of the iterations encounter overflow ?

Reply:
Yep. It seems to be the default value used in fairseq.